confgenerator/confgenerator_test.go

```diff
@@ -191,6 +191,17 @@ func updateOrCompareGolden(t *testing.T, testName string, goos string, expectedB
 		} else {
 			t.Errorf("test %q: golden file at %s mismatch (-want +got):\n%s", testName, goldenPath, diff)
 		}
+	} else {
+		if *updateGolden {
+			// If the diff is empty and we need to update goldens, check whether the diff is
+			// empty because the file doesn't exist yet.
+			if _, err := os.Stat(goldenPath); err != nil {
+				t.Logf("Detected -update_golden flag. Creating %q golden file.", goldenPath)
+				if err := ioutil.WriteFile(goldenPath, []byte(""), 0644); err != nil {
```

Comment:
Can we combine the code that writes the golden file from both branches? E.g., something like this:
```
	diff := cmp.Diff(expected, actual)
	if *updateGolden {
		if diff != "" || actual == "" {
			// Update the expected to match the actual.
			t.Logf("Detected -update_golden flag. Rewriting the %q golden file to apply the following diff\n%s.", goldenPath, cmp.Diff(actual, expected))
			if err := ioutil.WriteFile(goldenPath, []byte(actual), 0644); err != nil {
				t.Fatalf("error updating golden file at %q : %s", goldenPath, err)
			}
		}
	} else if diff != "" {
		t.Errorf("test %q: golden file at %s mismatch (-want +got):\n%s", testName, goldenPath, diff)
	}
```

Reply:
I got something similar to this to work; I modified it slightly from the example here because we only want it to write the file on empty diff if the file doesn't already exist. Let me know what you think.

Reply:
It's interesting that `readFileContent` assumes that any error from `ioutil.ReadFile` would be a "file not found" error. This means that if a non-empty file exists, but there's an intermittent error actually reading that file, the content will be assumed empty, and your code will not try to overwrite it, which would cause the tests to fail later (exactly what `-update_golden` is intended to prevent). If we want to catch that case, we should unconditionally attempt to overwrite the golden file when the content is empty (which will at least log a failure message in case of an intermittent error). Alternatively, we can fix line 168 to actually distinguish the various errors and only swallow the "file not found" ones… Does this make sense?